File: dotnetcoresignalrclientjava/src/main/java/com/smartarmenia/dotnetcoresignalrclientjava/WebSocketHubConnection.java
package com.smartarmenia.dotnetcoresignalrclientjava;

import android.net.Uri;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonElement;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft_6455;
import org.java_websocket.handshake.ServerHandshake;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class WebSocketHubConnection implements HubConnection {
    private static String SPECIAL_SYMBOL = "\u001E";
    private static String TAG = "WebSockets";

    private WebSocketClient client;
    private List<HubConnectionListener> listeners = new ArrayList<>();
    private Map<String, List<HubEventListener>> eventListeners = new HashMap<>();
    private Uri parsedUri;
    private String hubUrl;
    private Gson gson = new Gson();

    public WebSocketHubConnection(String hubUrl) {
        this.hubUrl = hubUrl;
        parsedUri = Uri.parse(hubUrl);
    }

    @Override
    public void connect(final String authHeader) {
        Runnable runnable = new Runnable() {
            public void run() {
                getConnectionId(authHeader);
            }
        };
        new Thread(runnable).start();
    }

    private void getConnectionId(String authHeader) {
        Log.i(TAG, "Requesting connection id...");
        if (!(parsedUri.getScheme().equals("http") || parsedUri.getScheme().equals("https")))
            throw new RuntimeException("URL must start with http or https");

        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(hubUrl).openConnection();
            if (authHeader != null && !authHeader.isEmpty()) {
                connection.addRequestProperty("Authorization", authHeader);
            }

            connection.setConnectTimeout(15000);
            connection.setReadTimeout(15000);
            connection.setRequestMethod("OPTIONS");
            int responseCode = connection.getResponseCode();

            if (responseCode == 200) {
                String result = InputStreamConverter.convert(connection.getInputStream());
                JsonElement jsonElement = gson.fromJson(result, JsonElement.class);
                String connectionId = jsonElement.getAsJsonObject().get("connectionId").getAsString();
                JsonElement availableTransportsElements = jsonElement.getAsJsonObject().get("availableTransports");
                List<String> availableTransports = Arrays.asList(gson.fromJson(availableTransportsElements, String[].class));
                if (!availableTransports.contains("WebSockets")) {
                    throw new RuntimeException("The server does not support WebSockets transport");
                }
                connectClient(connectionId, authHeader);
            } else if (responseCode == 401) {
                throw new RuntimeException("Unauthorized request");
            } else {
                throw new RuntimeException("Server error");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void connectClient(String connectionId, String authHeader) throws Exception {
        Uri.Builder uriBuilder = parsedUri.buildUpon();
        uriBuilder.appendQueryParameter("id", connectionId);
        uriBuilder.scheme(parsedUri.getScheme().replace("http", "ws"));
        Uri uri = uriBuilder.build();
        Map<String, String> headers = new HashMap<>();
        if (authHeader != null && !authHeader.isEmpty()) {
            headers.put("Authorization", authHeader);
        }
        client = new WebSocketClient(new URI(uri.toString()), new Draft_6455(), headers, 15000) {
            @Override
            public void onOpen(ServerHandshake handshakedata) {
                Log.i(TAG, "Opened");
                for (HubConnectionListener listener : listeners) {
                    listener.onConnected();
                }
                send("{\"protocol\":\"json\"}" + SPECIAL_SYMBOL);
            }

            @Override
            public void onMessage(String message) {
                Log.i(TAG, message);
                SignalRMessage element = gson.fromJson(message.replace(SPECIAL_SYMBOL, ""), SignalRMessage.class);
                if (element.getType() == 1) {
                    HubMessage hubMessage = new HubMessage(element.getInvocationId(), element.getTarget(), element.getArguments());
                    for (HubConnectionListener listener : listeners) {
                        listener.onMessage(hubMessage);
                    }

                    List<HubEventListener> hubEventListeners = eventListeners.get(hubMessage.getTarget());
                    if (hubEventListeners != null) {
                        for (HubEventListener listener : hubEventListeners) {
                            listener.onEventMessage(hubMessage);
                        }
                    }
                }
            }

            @Override
            public void onClose(int code, String reason, boolean remote) {
                Log.i(TAG, String.format("Closed. Code: %s, Reason: %s, Remote: %s", code, reason, remote));
                for (HubConnectionListener listener : listeners) {
                    listener.onDisconnected();
                }
            }

            @Override
            public void onError(Exception ex) {
                Log.i(TAG, "Error " + ex.getMessage());
                error(ex);
            }
        };
        Log.i(TAG, "Connecting...");
        client.connect();
    }

    private void error(Exception ex) {
        for (HubConnectionListener listener : listeners) {
            listener.onError(ex);
        }
    }

    @Override
    public void disconnect() {
        Runnable runnable = new Runnable() {
            public void run() {
                if (client != null && !(client.isClosed() || client.isClosing()))
                    client.close();
            }
        };
        new Thread(runnable).start();
    }

    @Override
    public void addListener(HubConnectionListener listener) {
        listeners.add(listener);
    }

    @Override
    public void removeListener(HubConnectionListener listener) {
        listeners.remove(listener);
    }

    @Override
    public void subscribeToEvent(String eventName, HubEventListener eventListener) {
        List<HubEventListener> eventMap;
        if (eventListeners.containsKey(eventName)) {
            eventMap = eventListeners.get(eventName);
        } else {
            eventMap = new ArrayList<>();
            eventListeners.put(eventName, eventMap);
        }
        eventMap.add(eventListener);
    }

    @Override
    public void unSubscribeFromEvent(String eventName, HubEventListener eventListener) {
        if (eventListeners.containsKey(eventName)) {
            List<HubEventListener> eventMap = eventListeners.get(eventName);
            eventMap.remove(eventListener);
            if (eventMap.isEmpty()) {
                eventListeners.remove(eventName);
            }
        }
    }

    @Override
    public void invoke(String event, Object... parameters) {
        if (client == null || !client.isOpen()) {
            throw new RuntimeException("Not connected");
        }
        final Map<String, Object> map = new HashMap<>();
        map.put("type", 1);
        map.put("invocationId", UUID.randomUUID());
        map.put("target", event);
        map.put("arguments", parameters);
        map.put("nonblocking", false);
        Runnable runnable = new Runnable() {
            public void run() {
                try {
                    client.send(gson.toJson(map) + SPECIAL_SYMBOL);
                } catch (Exception e) {
                    error(e);
                }
            }
        };
        new Thread(runnable).start();
    }

    private static class InputStreamConverter {
        static String convert(InputStream stream) throws IOException {
            BufferedReader r = new BufferedReader(new InputStreamReader(stream));
            StringBuilder total = new StringBuilder();
            String line;
            while ((line = r.readLine()) != null) {
                total.append(line);
                total.append('\n');
            }

            return total.toString();
        }
    }
}

File: dotnetcoresignalrclient/src/main/java/com/smartarmenia/dotnetcoresignalrclient/WebSocketHubConnection.kt
package com.smartarmenia.dotnetcoresignalrclient

import android.net.Uri
import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonElement
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft_6455
import org.java_websocket.handshake.ServerHandshake
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.*

class WebSocketHubConnection(private val hubUrl: String) : HubConnection {
    companion object {
        const val SPECIAL_SYMBOL = "\u001E"
        const val TAG = "WebSockets"
    }

    private lateinit var client: WebSocketClient
    private val listeners = mutableListOf<HubConnectionListener>()
    private val eventListeners = mutableMapOf<String, MutableList<HubEventListener>>()
    private val parsedUri = Uri.parse(hubUrl)
    private val gson = Gson()

    override fun connect(authHeader: String?) {
        Log.i(TAG, "Requesting connection id...")
        if (!(parsedUri.scheme == "http" || parsedUri.scheme == "https"))
            throw RuntimeException("URL must start with http or https")

        val connection: HttpURLConnection = URL(hubUrl).openConnection() as HttpURLConnection
        if (authHeader != null) {
            connection.addRequestProperty("Authorization", authHeader)
        }
        connection.connectTimeout = 15000
        connection.readTimeout = 15000
        connection.requestMethod = "OPTIONS"
        val responseCode = connection.responseCode
        when (responseCode) {
            200 -> {
                val result = InputStreamConverter.convert(connection.inputStream)
                val jsonElement = gson.fromJson<JsonElement>(result, JsonElement::class.java)
                val connectionId = jsonElement.asJsonObject.get("connectionId").asString
                if (!jsonElement.asJsonObject.get("availableTransports").asJsonArray.toList().map { it.asString }.contains("WebSockets")) {
                    throw RuntimeException("The server does not support WebSockets transport")
                }
                connectClient(connectionId, authHeader)
            }
            401 -> throw RuntimeException("Unauthorized request")
            else -> throw RuntimeException("Server error")
        }
    }

    private fun connectClient(connectionId: String?, authHeader: String?) {
        val uriBuilder = parsedUri.buildUpon()
        uriBuilder.appendQueryParameter("id", connectionId)
        uriBuilder.scheme(parsedUri.scheme.replace("http", "ws"))
        val uri = uriBuilder.build()
        val headers = if (authHeader == null) null else mapOf("Authorization" to authHeader)
        client = object : WebSocketClient(URI(uri.toString()), Draft_6455(), headers, 15000) {
            override fun onOpen(serverHandshake: ServerHandshake) {
                Log.i(TAG, "Opened")
                listeners.forEach { it.onConnected() }
                send("{\"protocol\":\"json\"}$SPECIAL_SYMBOL")
            }

            override fun onMessage(s: String) {
                Log.i(TAG, s)
                val element = gson.fromJson<SignalRMessage>(s.replace(SPECIAL_SYMBOL, ""), SignalRMessage::class.java)
                if (element.type == 1) {
                    val message = HubMessage(element.invocationId!!, element.target!!, element.arguments!!)
                    listeners.forEach { it.onMessage(message) }
                    eventListeners.get(message.target)?.forEach { it.onEventMessage(message) }
                }
            }

            override fun onClose(i: Int, s: String, b: Boolean) {
                Log.i(TAG, "Closed. Code: $i, Reason: $s, Remote: $b")
                listeners.forEach { it.onDisconnected() }
            }

            override fun onError(e: Exception) {
                Log.i(TAG, "Error " + e.message)
                listeners.forEach { it.onError(e) }
            }
        }
        Log.i(TAG, "Connecting...")
        client.connect()
    }

    override fun disconnect() {
        client.close()
    }

    override fun subscribeToEvent(eventName: String, eventListener: HubEventListener) {
        val eventMap: MutableList<HubEventListener>
        if (eventListeners.contains(eventName)) {
            eventMap = eventListeners.get(eventName)!!
        } else {
            eventMap = mutableListOf()
            eventListeners.put(eventName, eventMap)
        }
        eventMap.add(eventListener)
    }

    override fun unSubscribeFromEvent(eventName: String, eventListener: HubEventListener) {
        if (eventListeners.contains(eventName)) {
            val eventMap = eventListeners.get(eventName)!!
            eventMap.remove(eventListener)
            if (eventMap.isEmpty()) {
                eventListeners.remove(eventName)
            }
        }
    }

    override fun invoke(event: String, vararg parameters: Any) {
        val map = mapOf("type" to 1, "invocationId" to UUID.randomUUID().toString(), "target" to event, "arguments" to parameters, "nonblocking" to false)
        client.send(gson.toJson(map) + SPECIAL_SYMBOL)
    }

    override fun addListener(listener: HubConnectionListener) {
        listeners.add(listener)
    }

    override fun removeListener(listener: HubConnectionListener) {
        listeners.remove(listener)
    }

    private object InputStreamConverter {
        @Throws(IOException::class)
        internal fun convert(stream: InputStream): String {
            val r = BufferedReader(InputStreamReader(stream))
            val total = StringBuilder()
            r.lineSequence().forEach { line ->
                total.append(line)
                total.append('\n')
            }

            return total.toString()
        }
    }
}